src/Npgsql/PoolManager.cs

```diff
@@ -13,20 +11,45 @@ namespace Npgsql;
 /// </remarks>
 static class PoolManager
 {
-    internal static ConcurrentDictionary<string, NpgsqlDataSource> Pools { get; } = new();
+    static ConcurrentDictionary<string, NpgsqlDataSource> _pools = new();
+
+    internal static ConcurrentDictionary<string, NpgsqlDataSource> Pools => _pools;
 
     internal static void Clear(string connString)
     {
-        // TODO: Actually remove the pools from here, #3387 (but be careful of concurrency)
-        if (Pools.TryGetValue(connString, out var pool))
-            pool.Clear();
+        var pools = _pools;
+        if (!pools.TryRemove(connString, out var pool))
+                return;
+
+        pool.Dispose();
+
+        // The same NpgsqlDataSource instance may be referenced via different connection strings
+        // in our dictionary, so we have to make sure that we remove all occurrences
+#if NET5_0_OR_GREATER
+        foreach (var pair in pools)
+            if (ReferenceEquals(pair.Value, pool))
+                pools.TryRemove(pair);
+#else
+        foreach (var pair in pools.ToArray())
+            if (ReferenceEquals(pair.Value, pool))
+                // Slightly unsafer than the modern TryRemove() overload above
+                // since it doesn't check the value for equality.
+                if (pools.TryRemove(pair.Key, out var otherPool) && !ReferenceEquals(otherPool, pool))
+                    // if got out a different pool, quickly
+                    // try to push it back and hope nobody notices it
+                    if (!pools.TryAdd(pair.Key, otherPool))
+                        // If somebody has noticed
+                        // at least clean up our mistake
+                        otherPool.Dispose();
+#endif
     }
 
     internal static void ClearAll()
     {
-        // TODO: Actually remove the pools from here, #3387 (but be careful of concurrency)
-        foreach (var pool in Pools.Values)
-            pool.Clear();
+        var pools = _pools;
+        _pools = new();
+        foreach (var pool in pools.Values)
```

Comment:
Unless I'm mistaken, if this is racing with Pools.GetOrAdd() from NpgsqlConnection.SetupDataSource(), a data source could get added which never gets disposed.

We may simply want to take a write lock on PoolManager whenever we want to mutate the dictionary (e.g. add a new pool, clear or whatever). Lookups (the thing that needs to be fast) could continue to be lock-free, with ConcurrentDictionary as today.

(note that the write lock would only be needed in ClearAll around the copying and replacing of the _pools reference. Just enough to make sure that nobody references the old dictionary when adding a new pool etc.)

Reply:
> Unless I'm mistaken, if this is racing with Pools.GetOrAdd() from NpgsqlConnection.SetupDataSource(), a data source could get added which never gets disposed.

Yes. It is indeed a tiny chance for a pool leak but given the fact that we come from a situation where we don't care about "leaking" all pools in 100% of the cases I considered it as negligible in favor of not locking anything.

> We may simply want to take a write lock on PoolManager whenever we want to mutate the dictionary (e.g. add a new pool, clear or whatever). Lookups (the thing that needs to be fast) could continue to be lock-free, with ConcurrentDictionary as today.
> 
> (note that the write lock would only be needed in ClearAll around the copying and replacing of the _pools reference. Just enough to make sure that nobody references the old dictionary when adding a new pool etc.)

I'm not sure I understand correctly. If we want to eliminate the aforementioned race condition, `GetOrAdd()` (or rather a method wrapping it) would also need a lock - no?
So, in that case only the initial two `TryGetValue()` calls in `SetupDataSource()` would stay in the optimized (non-locked) path.
Or are you talking about `ReaderWriterLockSlim` again?
I'm asking because I'm not sure, if the lock is contended enough to justify that overhead.

Reply:
Ok, I've implemented it the way I think might be best to avoid any leaking of pools.
Let me know what you think about it.

Reply:
Yeah, what you've implemented is what I had in mind (modulu the missing lock taking I commented).

My reasoning here is that locking only for *updates* of the dictionary should be fine, since that's very rare; the high-perf path (where we should avoid contention) is only when reading that dictionary (normal usage). Since the pool really is disposable (it manages a timer internally), I do think it's important that correct user code with Clear/ClearAll not leak pools in general, otherwise lots of timers could accumulate in the application. I agree it's an edge case, but handling it also doesn't seem super difficult (as you've done).